### PR TITLE
Masking specs

### PR DESCRIPTION
**Pull request by fmalita**
Schema and docs for layer masks.

### CODE REVIEW

**Comment by fmalita** [comment]
Bodymovin also sets a `hasMask` bool for each layer, but that seems redundant vs. just checking for `masksProperties`. I propose we leave it out,  wdyt?

**Comment by b-wils** [comment]
> Bodymovin also sets a `hasMask` bool for each layer, but that seems redundant vs. just checking for `masksProperties`. I propose we leave it out, wdyt?

That makes sense to me, though at least lottie-web references it so that may need to be updated (with the usual compat concern for existing lotties)- https://github.com/search?q=repo%3Aairbnb%2Flottie-web%20hasmask&type=code

**Comment by fmalita** [comment]
Ah, good point, existing players probably depend on it.  Maybe it's more prudent to add it initially, and shoot for removal in a later compat-breaking revision.

File: docs/specs/helpers.md [hunk 1]
```diff
@@ -56,3 +56,52 @@ The anchor point is highlighted with an orange dot.
...
+        <input type="range" min="0" max="100" value="100" title="Opacity"/>
+        <input type="range" min="-40" max="40" value="0" title="Expansion"/>
+        <input type="checkbox" title="Invert"/>
+    </form>
+    <json>lottie.layers[1].masksProperties[0]</json>
+    <script>
+        let mask = lottie.layers[1].masksProperties[0];
+        mask.o.k = Number(data["Opacity"]);
+        mask.x.k = Number(data["Expansion"]);
+        mask.inv = data["Invert"];
```
Comment (b-wils) [inline docs/specs/helpers.md:104]: It looks like invert doesn't work properly in LottieWeb player. I'm not sure what the best option is here-

1. Remove field from playground (and maybe spec???).
2. Fix LottieWeb.
3. Use a different player (Skottie?).
4. Leave as is and document player issue/hope there is no confusion.

File: schema/helpers/mask.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "type": "object",
+    "title": "Mask",
+    "description": "Mask for layer content.",
+    "allOf": [
+        {
+            "properties": {
```
Comment (b-wils) [inline schema/helpers/mask.json:8]: Should some of these properties be required? I think `pt` for sure. The rest seem like though could be optional, though I wonder if players will handle that.

File: schema/layers/visual-layer.json [hunk 3]
```diff
@@ -19,7 +19,20 @@
                     "title": "Auto Orient",
                     "$ref": "#/$defs/values/int-boolean",
                     "default": 0,
-                    "description": "If 1, The layer will rotate itself to match its animated position path"
+                    "description": "If 1, the layer will rotate itself to match its animated position path"
+                },
+                "hasMask": {
```
Comment (b-wils) [inline schema/layers/visual-layer.json:24]: Consider adding a conditional to ensure hasMask and masksProperties are in sync (either both set or both unset).

File: schema/helpers/mask.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+                "o": {
+                    "title": "Opacity",
+                    "$ref": "#/$defs/properties/scalar-property"
+                },
+                "pt": {
+                    "title": "Shape",
+                    "$ref": "#/$defs/properties/bezier-property"
+                },
+                "x": {
+                    "title": "Expansion",
```
Comment (b-wils) [inline schema/helpers/mask.json:28]: Add description for expansion? Something about how this is processed would be useful too.

**Review by b-wils — CHANGES_REQUESTED** [review]

File: docs/specs/helpers.md (hunk 1 above)
Comment (mbasaglia) [reply docs/specs/helpers.md:104]: IMO we should not include in specs 1.0 features that are not widely supported by existing players

File: schema/layers/visual-layer.json (hunk 3 above)
Comment (mbasaglia) [reply schema/layers/visual-layer.json:24]: When we discussed this we said we would not include `hasMasks` in the specs

File: docs/specs/helpers.md (hunk 1 above)
Comment (fmalita) [reply docs/specs/helpers.md:104]: Are you thinking about this as reducing the scope for 1.0 specifically, or more of a general direction going forward?  Reducing 1.0 scope sounds reasonable, but if the latter, we should try to clarify a few points:

- isn't player capability/support supposed to be covered by profiles?
- what is the actual player set we are considering when making decisions?
- is the feature intersection going to be too restrictive?
- in practice, how do we determine support for players not actively represented?
- support vs implementation bugs - e.g. inverted masks appear to be supported in LottieWeb, they just produce incorrect results (in some cases/backends?)
- not all players are developed/supported at the same pace - what are the implications for long term format evolution?

For the issue at hand, I think I'm fine with dropping inverted masks (and mask expansion, and the uncommon mask modes) from 1.0, but IMO they should be added to the spec in future revisions even if not supported by all players.

Comment (mbasaglia) [reply docs/specs/helpers.md:104]: (Repeating here for visibility) I think just for 1.0 we should have features that are supported by most players and worry about profiles etc. after we published the first version of the specs

File: docs/specs/helpers.md [hunk 5]
```diff
@@ -56,3 +56,52 @@ The anchor point is highlighted with an orange dot.
...
+
+{schema_string:helpers/mask/description}
+
+{schema_object:helpers/mask}
+
+Masks provide single-channel coverage information (alpha channel) that modulates the layer's
+content.
+
+When multiple masks are specified, they are combined (blended) into a single coverage buffer,
+in order, based on the [`mode`](../constants/#mask-mode) operator.
```
[the remark this answers is not in the document]
Comment (fmalita) [reply docs/specs/helpers.md]: Updated.

File: schema/helpers/mask.json (hunk 2 above)
Comment (fmalita) [reply schema/helpers/mask.json:8]: Required `pt` makes sense, added.

File: schema/helpers/mask.json (hunk 4 above)
Comment (fmalita) [reply schema/helpers/mask.json]: Dropped expansion, it is poorly supported across players.

File: docs/specs/helpers.md (hunk 1 above)
Comment (fmalita) [reply docs/specs/helpers.md]: Sounds good, thanks.  I dropped the inverted option, expansion, and the uncommon/unsupported mask modes.

File: schema/layers/visual-layer.json (hunk 3 above)
Comment (fmalita) [reply schema/layers/visual-layer.json]: Removed `hasMasks`.

**Review by b-wils — APPROVED** [review]

**Review by mbasaglia — APPROVED** [review]

File: schema/helpers/mask.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    "type": "object",
+    "title": "Mask",
+    "description": "Mask for layer content.",
+    "allOf": [
+        {
+            "properties": {
+                "mode": {
+                    "title": "Mode",
+                    "$ref": "#/$defs/constants/mask-mode",
+                    "default": "a"
```
Comment (Aidosmf) [inline schema/helpers/mask.json:12]: [lottie-docs/schema](https://lottiefiles.github.io/lottie-docs/schema/#/$defs/helpers/mask) says the default mask mode is `i`, are we ok changing it?

```suggestion
                    "default": "i"
```

File: schema/helpers/mask.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+                },
+                "o": {
+                    "title": "Opacity",
+                    "description": "Mask opacity, as a percentage [0..100].",
+                    "$ref": "#/$defs/properties/scalar-property",
+                    "default": 100
+                },
+                "pt": {
+                    "title": "Shape",
+                    "description": "Mask shape.",
```
Comment (Aidosmf) [inline schema/helpers/mask.json:22]:
```suggestion
                    "description": "Mask shape",
```

**Review by Aidosmf — CHANGES_REQUESTED** [review]

File: schema/helpers/mask.json (hunk 6 above)
Comment (fmalita) [reply schema/helpers/mask.json:12]: Not sure where the `i` default is coming from, in AE the default mode for new masks is Add.

That said, BM always exports an explicit mode, and intersection by default sort of makes sense from a graphics API point or view - so keeping `i` sgtm.

**Review by Aidosmf — APPROVED** [review]

**Pull request merged**